<center><span style="color:#b30000;font-size:35px;"><br><strong>Analysis Phase</strong></span></center>

<span style="color:#2929a3;font-size:20px;">Import Libraries</span>

In [70]:
import pandas as pd
import numpy as np
import plotly.express as px

<span style="color:#2929a3;font-size:20px;">Read Dataset</span>

In [71]:
df = pd.read_pickle('Data/df_RFM_final2.pkl')
df.head()

Mer_Name  Recency  Frequency  Monetary  Cluster
User_Id                                                                  
0          Jai Hind General Stores      281          1      4427        0
1             Agra Appliance Arena       34          1      2490        2
2        Amritsar Auto Accessories      236          1      3000        2
3             Agra Appliance Arena      320          1      1620        2
4                Ranchi Rug Retail      407          1       700        0

<span style="color:#b30000;font-size:30px;"><strong>Analyze each Cluster </strong></span>

<span style="color:#b30000;font-size:25px;">Cluster 0 : </span>

In [72]:
df_cls0 = df[df['Cluster'] == 0]

In [73]:
df_cls0['Mer_Name'].value_counts()

Mer_Name
Rajasthan Handicrafts      15936
Jai Hind General Stores     2688
Faridabad Footwear Fair      299
Ranchi Rug Retail            101
Name: count, dtype: int64

In [74]:
## List of merchants exist in cluster 0
merchants_cls0_list = df_cls0['Mer_Name'].unique().tolist()

<span style="color:#2929a3;font-size:19px;">Recency of each Merchant </span>

In [75]:
## Show distribution of "Recency" in each "Merchant" in "cluster0"
for merchant in merchants_cls0_list:
    df_merchant = df_cls0[df_cls0['Mer_Name'] == merchant]
    fig = px.histogram(data_frame=df_merchant, x='Recency', title='Recency of "'+merchant+'" Merchant', text_auto=True)
    fig.show()

<span style="color:#2929a3;font-size:19px;">Frequency of each Merchant </span>

In [76]:
## Show distribution of "Frequency" in each "Merchant" in "cluster0"
for merchant in merchants_cls0_list:
    df_merchant = df_cls0[df_cls0['Mer_Name'] == merchant]
    fig = px.histogram(data_frame=df_merchant, x='Frequency', title='Frequency of "'+merchant+'" Merchant', text_auto=True)
    fig.show()

<span style="color:#2929a3;font-size:19px;">Monetary of each Merchant </span>

In [77]:
## Show distribution of "Monetary" in each "Merchant" in "cluster0"
for merchant in merchants_cls0_list:
    df_merchant = df_cls0[df_cls0['Mer_Name'] == merchant]
    fig = px.histogram(data_frame=df_merchant, x='Monetary', title='Monetary of "'+merchant+'" Merchant', text_auto=True)
    fig.show()

In [78]:
## Show Median Value of (RFM) of each merchant
for merchant in merchants_cls0_list:
    df_merchant = df_cls0[df_cls0['Mer_Name'] == merchant]
    print(f'Median Recency of "{merchant}" merchant is ',df_merchant['Recency'].median())
    print(f'Median Frequency of "{merchant}" merchant is ',df_merchant['Frequency'].median())
    print(f'Median Monetary of "{merchant}" merchant is ',df_merchant['Monetary'].median())
    print("*" * 70)

Median Recency of "Jai Hind General Stores" merchant is  266.0
Median Frequency of "Jai Hind General Stores" merchant is  1.0
Median Monetary of "Jai Hind General Stores" merchant is  2000.0
**********************************************************************
Median Recency of "Ranchi Rug Retail" merchant is  295.0
Median Frequency of "Ranchi Rug Retail" merchant is  1.0
Median Monetary of "Ranchi Rug Retail" merchant is  1360.0
**********************************************************************
Median Recency of "Rajasthan Handicrafts" merchant is  90.0
Median Frequency of "Rajasthan Handicrafts" merchant is  1.0
Median Monetary of "Rajasthan Handicrafts" merchant is  2140.0
**********************************************************************
Median Recency of "Faridabad Footwear Fair" merchant is  34.0
Median Frequency of "Faridabad Footwear Fair" merchant is  1.0
Median Monetary of "Faridabad Footwear Fair" merchant is  2960.0
*************************************************

### Top 3 Merchants in Cluster 0:
#### 1 - Faridabad Footwear Fair
#### 2 - Rajasthan Handicrafts
#### 3 - Jai Hind General Stores

<span style="color:#b30000;font-size:25px;">Cluster 1 : </span>

In [79]:
df_cls1 = df[df['Cluster'] == 1]

In [80]:
df_cls1['Mer_Name'].value_counts()

Mer_Name
Madurai Music Mania         3253
Indore Instrument Inn       1151
Mumbai Fashion Hub           682
Pune Perfumes Paradise       437
Varanasi Silk House          147
Jodhpur Jewelry Junction      86
Name: count, dtype: int64

In [81]:
## List of merchants exist in cluster 1
merchants_cls1_list = df_cls1['Mer_Name'].unique().tolist()

<span style="color:#2929a3;font-size:19px;">Recency of each Merchant </span>

In [82]:
## Show distribution of "Recency" in each "Merchant" in "cluster1"
for merchant in merchants_cls1_list:
    df_merchant = df_cls1[df_cls1['Mer_Name'] == merchant]
    fig = px.histogram(data_frame=df_merchant, x='Recency', title='Recency of "'+merchant+'" Merchant', text_auto=True)
    fig.show()

<span style="color:#2929a3;font-size:19px;">Frequency of each Merchant </span>

In [83]:
## Show distribution of "Frequency" in each "Merchant" in "cluster1"
for merchant in merchants_cls1_list:
    df_merchant = df_cls1[df_cls1['Mer_Name'] == merchant]
    fig = px.histogram(data_frame=df_merchant, x='Frequency', title='Frequency of "'+merchant+'" Merchant', text_auto=True)
    fig.show()

<span style="color:#2929a3;font-size:19px;">Monetary of each Merchant </span>

In [84]:
## Show distribution of "Monetary" in each "Merchant" in "cluster1"
for merchant in merchants_cls1_list:
    df_merchant = df_cls1[df_cls1['Mer_Name'] == merchant]
    fig = px.histogram(data_frame=df_merchant, x='Monetary', title='Monetary of "'+merchant+'" Merchant', text_auto=True)
    fig.show()

In [85]:
## Show Median Value of (RFM) of each merchant
for merchant in merchants_cls1_list:
    df_merchant = df_cls1[df_cls1['Mer_Name'] == merchant]
    print(f'Median Recency of "{merchant}" merchant is ',df_merchant['Recency'].median())
    print(f'Median Frequency of "{merchant}" merchant is ',df_merchant['Frequency'].median())
    print(f'Median Monetary of "{merchant}" merchant is ',df_merchant['Monetary'].median())
    print("*" * 70)

Median Recency of "Madurai Music Mania" merchant is  226.0
Median Frequency of "Madurai Music Mania" merchant is  1.0
Median Monetary of "Madurai Music Mania" merchant is  2069.0
**********************************************************************
Median Recency of "Indore Instrument Inn" merchant is  222.0
Median Frequency of "Indore Instrument Inn" merchant is  1.0
Median Monetary of "Indore Instrument Inn" merchant is  1940.0
**********************************************************************
Median Recency of "Pune Perfumes Paradise" merchant is  245.0
Median Frequency of "Pune Perfumes Paradise" merchant is  1.0
Median Monetary of "Pune Perfumes Paradise" merchant is  1999.0
**********************************************************************
Median Recency of "Mumbai Fashion Hub" merchant is  250.0
Median Frequency of "Mumbai Fashion Hub" merchant is  1.0
Median Monetary of "Mumbai Fashion Hub" merchant is  2880.0
***********************************************************

### Top 3 Merchants in Cluster 1:
#### 1 - Madurai Music Mania
#### 2 - Indore Instrument Inn
#### 3 - Mumbai Fashion Hub

<span style="color:#b30000;font-size:25px;">Cluster 2 : </span>

In [86]:
df_cls2 = df[df['Cluster'] == 2]

In [87]:
df_cls2['Mer_Name'].value_counts()

Mer_Name
Saraswati Fabrics            6184
Agra Appliance Arena         2568
Amritsar Auto Accessories    1190
Gurgaon Gift Gallery          873
Krishna Textiles              138
Gwalior Garden Gear           122
Name: count, dtype: int64

In [88]:
## List of merchants exist in cluster 2
merchants_cls2_list = df_cls2['Mer_Name'].unique().tolist()

<span style="color:#2929a3;font-size:19px;">Recency of each Merchant </span>

In [89]:
## Show distribution of "Recency" in each "Merchant" in "cluster2"
for merchant in merchants_cls2_list:
    df_merchant = df_cls2[df_cls2['Mer_Name'] == merchant]
    fig = px.histogram(data_frame=df_merchant, x='Recency', title='Recency of "'+merchant+'" Merchant', text_auto=True)
    fig.show()

<span style="color:#2929a3;font-size:19px;">Frequency of each Merchant </span>

In [90]:
## Show distribution of "Frequency" in each "Merchant" in "cluster2"
for merchant in merchants_cls2_list:
    df_merchant = df_cls2[df_cls2['Mer_Name'] == merchant]
    fig = px.histogram(data_frame=df_merchant, x='Frequency', title='Frequency of "'+merchant+'" Merchant', text_auto=True)
    fig.show()

<span style="color:#2929a3;font-size:19px;">Monetary of each Merchant </span>

In [91]:
## Show distribution of "Monetary" in each "Merchant" in "cluster2"
for merchant in merchants_cls2_list:
    df_merchant = df_cls2[df_cls2['Mer_Name'] == merchant]
    fig = px.histogram(data_frame=df_merchant, x='Monetary', title='Monetary of "'+merchant+'" Merchant', text_auto=True)
    fig.show()

In [92]:
## Show Median Value of (RFM) of each merchant
for merchant in merchants_cls2_list:
    df_merchant = df_cls2[df_cls2['Mer_Name'] == merchant]
    print(f'Median Recency of "{merchant}" merchant is ',df_merchant['Recency'].median())
    print(f'Median Frequency of "{merchant}" merchant is ',df_merchant['Frequency'].median())
    print(f'Median Monetary of "{merchant}" merchant is ',df_merchant['Monetary'].median())
    print("*" * 70)

Median Recency of "Agra Appliance Arena" merchant is  182.5
Median Frequency of "Agra Appliance Arena" merchant is  1.0
Median Monetary of "Agra Appliance Arena" merchant is  2500.0
**********************************************************************
Median Recency of "Amritsar Auto Accessories" merchant is  89.0
Median Frequency of "Amritsar Auto Accessories" merchant is  1.0
Median Monetary of "Amritsar Auto Accessories" merchant is  1860.0
**********************************************************************
Median Recency of "Saraswati Fabrics" merchant is  236.0
Median Frequency of "Saraswati Fabrics" merchant is  1.0
Median Monetary of "Saraswati Fabrics" merchant is  1868.5
**********************************************************************
Median Recency of "Gurgaon Gift Gallery" merchant is  270.0
Median Frequency of "Gurgaon Gift Gallery" merchant is  1.0
Median Monetary of "Gurgaon Gift Gallery" merchant is  1800.0
******************************************************

In [93]:
df_cls2['Mer_Name'].value_counts()

Mer_Name
Saraswati Fabrics            6184
Agra Appliance Arena         2568
Amritsar Auto Accessories    1190
Gurgaon Gift Gallery          873
Krishna Textiles              138
Gwalior Garden Gear           122
Name: count, dtype: int64

Saraswati Fabrics            6184
Agra Appliance Arena         2568
Amritsar Auto Accessories    1190
Gurgaon Gift Gallery          873
Krishna Textiles              138
Gwalior Garden Gear           122

### Top 3 Merchants in Cluster 2:
#### 1 - Amritsar Auto Accessories
#### 2 - Agra Appliance Arena
#### 3 - Saraswati Fabrics

<span style="color:#b30000;font-size:30px;"><strong>Summary of Analysis </strong></span>

### Top 3 Merchants in Cluster 0:
#### 1 - Faridabad Footwear Fair
#### 2 - Rajasthan Handicrafts
#### 3 - Jai Hind General Stores

#### ##################################################################
### Top 3 Merchants in Cluster 1:
#### 1 - Madurai Music Mania
#### 2 - Indore Instrument Inn
#### 3 - Mumbai Fashion Hub

#### ##################################################################
### Top 3 Merchants in Cluster 2:
#### 1 - Amritsar Auto Accessories
#### 2 - Agra Appliance Arena
#### 3 - Saraswati Fabrics

<span style="color:#2929a3;font-size:19px;">Dump Merchants_dict</span>

In [157]:
## Create Merchants dict
Merchants_dict = {
    0:['Faridabad Footwear Fair','Rajasthan Handicrafts','Jai Hind General Stores'],
    1:['Madurai Music Mania','Indore Instrument Inn','Mumbai Fashion Hub'],
    2:['Amritsar Auto Accessories','Agra Appliance Arena','Saraswati Fabrics']
}

In [158]:
import joblib 
joblib.dump(Merchants_dict, 'Data/Merchants_dict.pkl')

['Data/Merchants_dict.pkl']

<span style="color:#b30000;font-size:25px;"><strong>Deployment Code </strong></span>

In [187]:
%%writefile Deployment.py

import pandas as pd
import numpy as np
import streamlit as st
import joblib

df = pd.read_pickle('Data/df_RFM_final2.pkl')
Merchants_dict = joblib.load('Data/Merchants_dict.pkl')

def Top_3_Merchants(user_id):
    
    ## In Case that User exist in one Cluster
    if np.size(np.unique(df.loc[user_id]['Cluster'])) == 1:
        return Merchants_dict[df.loc[user_id]['Cluster']]
    ## In Case that User exist in the 2 Clusters
    elif np.size(np.unique(df.loc[user_id]['Cluster'])) == 2:
        merchant_list = []
        merchant_list.append(Merchants_dict[np.unique(df.loc[user_id]['Cluster'].values)[0]][0]) ## First Merchant in First cluster
        merchant_list.append(Merchants_dict[np.unique(df.loc[user_id]['Cluster'].values)[0]][1]) ## Second Merchant in First Cluster 
        merchant_list.append(Merchants_dict[np.unique(df.loc[user_id]['Cluster'].values)[1]][0]) ## First Merchant in Second Cluster
        return merchant_list
    ## In Case that User exist in the 3 Clusters
    else:
        merchant_list = []
        merchant_list.append(Merchants_dict[np.unique(df.loc[user_id]['Cluster'].values)[0]][0]) ## First Merchant in First Cluster
        merchant_list.append(Merchants_dict[np.unique(df.loc[user_id]['Cluster'].values)[1]][0]) ## First Merchant in Second Cluster
        merchant_list.append(Merchants_dict[np.unique(df.loc[user_id]['Cluster'].values)[2]][0]) ## First Merchant in Third Cluster
        return merchant_list
            
        
def Main():
    st.title('Customer Segmentation')
    user_id = st.number_input("User ID: ",format="%d", step=1, value=0)
    if st.button('Show Merchants'):
        if user_id in df.index:
            merchants = Top_3_Merchants(user_id)
            st.text(f'The user can spend his/her points in : \n 1- {merchants[0]}\n 2- {merchants[1]}\n 3- {merchants[2]}')
        else:
            st.text("User ID doesn't exist !!")
    
Main()

Overwriting Deployment.py
